### ISSUE TITLE
Add value-returning logical-OR, i.e. a false-coalescing operator

### ISSUE THREAD

**Issue by bilbothebaggins** [PowerShell/PowerShell#24087]
## Prelude

Powershell's logical operators (`-or` `-and` etc.) return a Boolean.

While Powershell already supports the [`$null ??` coalescing operator](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.4#null-coalescing-operator-) - 
**The problem with Powershell is, that more often than not, I want to treat $null and the empty string the same way.**
See e.g. this question: https://github.com/PowerShell/PowerShell/issues/21410

As a reference, in Perl the logical operators do NOT return boolean, but the left or right hand side VALUE: https://perldoc.perl.org/perlop#C-style-Logical-Or
```
$first_value = $maybe_empty_or_value || $another || $yet_another || get_stuff_eval_once() || ...;
```
That way it is very easy to assign the first of several alternatives.

## Proposal

I propose to add operators `-vOr` (and possibly `-vAnd`) (or whatever we'd want to call those) that do *not return boolean* but instead the left or right **value**, so that I could write:

```
$firstFound = $(Get-OptionalExpensiveStr $query) -vOR "not found"

# instead of

$firstFound = Get-OptionalExpensiveStr $query
if (-not $firstFound) {
  $firstFound = "not found"
}

```

For $null, the `-vOR` operator would be equivalent to the null coalescing operator - the usefulness of which has already been acknowledged I think.

----

## Aside:

For a strongly typed language like C#, extending the ternary operator `[bool] ? "yes" : "no"` into a **null** _coalescing_ operator `maybe ?? shmaybe ?? "definitely"` made sense insofar, as the first ternary term always must be a strongly typed boolean, so _coalescing_ more or less only made sense wrt. `null`.

For a dynamic language like Powershell, where _**any** value can be tested_, it would IMHO really have been better not to take over the _null_ coalescing behaviour, but have the `??` operator behave like I propose for `-vOR` above - like OP expected in #21410 
I guess that ship has sailed: So I get to my proposal above :-)

**Comment by bilbothebaggins** [comment]
@237dmitry 

> > For $null, the -vOR operator would be equivalent to the null coalescing operator
> 
> Replace `??=` with `-vOR` ?
> ...

The [Null-coalescing operator `??`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.4#null-coalescing-operator-) is what I was talking about.

You posted the [Null-coalescing assignment operator `??=`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.4#null-coalescing-assignment-operator-)

Otherwise, sorry, I'm not sure what you're meaning to say.
The Null-coalescing assignment operator has the same null-vs-empty problem as the other one?

**Comment by IISResetMe** [comment]
You could achieve the same by wrapping each operand expression in a scriptblock and just testing the produced value until you get a non-falsy value:

```pwsh
function Get-FirstTruthy {
  param([scriptblock[]]$ValueGenerators)
  
  foreach($valueGenerator in $ValueGenerators) {
    if ($generatedValue = &$valueGenerator) {
      return $generatedValue
    }
  }
}

$first = Get-FirstTruthy {$maybe_empty_or_value},{$another},{$yet_another},{Get-StuffEvalOnce}
```

**Comment by 237dmitry** [comment]
> I'm not sure what you're meaning to say

Like your perl example:

```powershell
$c = 1
$d ??= $a ??= $b ??= $c
$d
```

**Comment by 237dmitry** [comment]
> only continues on encountering $null on the right-hand side

Yes, variable shows/gets the first non-null value from the chained `?? / ??=`:

![ArcoLinux_2024-08-01_18-24-13](https://github.com/user-attachments/assets/34a721a4-2d71-46d2-b6a3-e05a3d932b72)

**Comment by IISResetMe** [comment]
@237dmitry Yes, but _`$false` is not the same as `$null`_. `$false ?? $false ?? $true` evaluates to `$false`, not `$true`

**Comment by 237dmitry** [comment]
Then you can use switch....

```powershell
switch ($true)
{
      $false {'a'}
      $true  {'b'}
      $false {'c'}
}
```

**Comment by IISResetMe** [comment]
@237dmitry I'm not sure I follow, can you elaborate? Where does `$a`/`$b`/`$c`/`$d` go in your `switch`?

**Comment by 237dmitry** [comment]
> Where does $a/$b/$c/$d go in your switch?

```powershell
$a = $false   
$b = $true        
$c = $true        
                  
switch ($true)    
{                 
      $a      {'a'; break}
      $b      {'b'; break}
      $c      {'c'}
      default {'d'}
}
```

**Comment by IISResetMe** [comment]
But that's not what OP is asking for either. They don't want exact mapping, they just want to shortcircuit evaluation of 2 or more expressions until a _truthy_ value emerges. In your example, if we set `$b = "truthy value that's not a boolean"` then that's the value we should be resolving. That doesn't work with your switch

**Comment by 237dmitry** [comment]
> But that's not what OP is asking

I doubt this is even necessary. Compare with the usage of ternary and null-coalescing operators, are they often used? I see this very rarely. And this takes into account that they were added quite a long time ago and yet are still not popular.

But I don’t mind, let’s have more sugar.

**Comment by bilbothebaggins** [comment]
> > But that's not what OP is asking
> 
> I doubt this is even necessary. Compare with the usage of ternary and null-coalescing operators, are they often used? I see this very rarely. 
> 

I see the null-coalescing operator heavily used *in C# code*.

My conjecture in the initial suggestion 

> powershell ... more often than not, I want to treat `$null` and the empty string the same way.

was that it may actually not be used because distinguishing `$null` from `$false` from `""` is not that useful in pwsh.
